service-base/src/main/java/org/eclipse/hono/service/AbstractProtocolAdapterBase.java

```diff
@@ -483,16 +485,24 @@ protected void doStop(final Future<Void> stopFuture) {
         final Handler<ProtonConnection> disconnectHandler = getHandlerForDisconnectHonoService(client, serviceName,
                 connectionEstablishedHandler, connectionLostHandler);
 
+        updateReadiness(procedureNameConnectedTo(serviceName), Status.KO()); // initially not ready
```

Comment:
I agree, I was only wondering if we need to explicitly report the KO value in order for this to be the case? Wouldn't the readiness check fail anyways if no result has been reported yet?

Reply:
> Wouldn't the readiness check fail anyways if no result has been reported yet?

No, not with the way the checks are used here. The mechanism implemented in this PR is that verticles perform checks and send the results via `EventBus` to `HealthCheckServer` which then registers a `procedure` (which essentially is a name and a function) at a Vert.X `HealthCheckHandler`. (Note that in our case the check function does not perform a check, but just returns the outcome of the check reported).
The consequence is, that `HealthCheckServer` does not know about the check and does not register a procure until it receives the first "update". 
In other words there is no difference between "registering" a check and "updating" it.
I hope this explanation is helpful and not even more confusing :-)

Reply:
So, you are saying that the health check for a service succeeds until reported otherwise by any of the verticles etc?

Reply:
Yes, that is probably the shortest way to put it :-)
When no check is registered on a `HealthCheckHandler` the liveness probe (e.g. http://127.0.0.1:8088/readiness) returns `204 No Content` and for Kubernetes the HTTP codes >=200 < 400 indicate success.

Reply:
Hm, lets try to walk through this: assume that the service is deployed to Kubernetes and run in a pod with a readiness check configured. When the pod starts and Kubernetes invokes the readiness check URL for the first time, then the result may or may not reflect the real status of the pod because it will depend on whether all verticles have already reported their initial `KO` status or not, right? If they have, then the result of the check will reflect the real status (KO), however, if at least one verticle hasn't, then the result might be correct or not, depending on whether the *missing* verticles are already OK or KO.

The problem seems to be that the health check server has no clue of how many verticles will be  contributing to the overall status, right?

Reply:
You are right. Since we are already digging deep: there is the additional option in Kubernetes to configure `initialDelaySeconds` on a health check, which is probably for such cases (if a `readinessProbe` is configured, the ["default state of readiness before the initial delay is `Failure`"](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-probes)).

> The problem seems to be that the health check server has no clue of how many verticles will be contributing to the overall status, right?

Exactly. The price for decoupling the health check server from the verticals is, that the former one does not know which verticles exist, how many instances have been deployed and which checks (procedures) they perform, _until_ they have reported them for the first time.

Regarding the concrete piece of code here, I still think it correct from the point of view of the Protocol Adapter to start reporting that it is not ready as early as possible and then change the state to ready as soon as it is connected to the required services.